pkg/dab/find.go

```diff
@@ -0,0 +1,237 @@
+//  - warpforge-error-searching-filesystem -- when the search of the filesystem produces an invalid result
+//  - warpforge-error-serialization -- when IPLD deserialization fails
+func FindActionableFromFS(
+	fsys fs.FS,
+	basisPath string, searchPath string, searchUp bool,
+	accept ActionableSearch,
+) (
+	m *wfapi.Module, p *wfapi.Plot, f *wfapi.Formula,
+	foundPath string, remainingSearchPath string, err error,
+) {
+	remainingSearchPath = searchPath
+
+	// First round: this may be a file, and may be one of any of the types.
+	// Stat it and consider that first:
+	// if it is a file, that's its own whole detection procedure (and means no further search);
+	// if it's not a file, we can procede to the behavior for dir feature detection (without popping).
+	foundPath = filepath.Join(basisPath, remainingSearchPath)
+	fi, e2 := stat(fsys, foundPath)
+	if e2 != nil {
+		err = e2
+		return
+	}
+	if fi.Mode()&^fs.ModePerm == 0 {
+		fh, e2 := open(fsys, foundPath)
+		defer fh.Close()
+		if e2 != nil {
+			err = e2
+			return
+		}
+		m, p, f, err = findActionableFromFile(fh, accept)
+		if m != nil { // If a module was found, a plot may also exist as a sibling file.
+			p, e2 = PlotFromFile(fsys, filepath.Join(filepath.Dir(filepath.Join(basisPath, remainingSearchPath)), "plot.wf"))
+			if !errors.Is(e2, fs.ErrNotExist) {
+				err = wfapi.ErrorSearchingFilesystem("loading plot associated with a module", e2)
+			}
+		}
+		return
+	}
+
+	// Iteratively check directories.
+	// In each directory, look for well-known file names.
+	// If we find a module, that dominates; then plot; then formula.
+	// When there's no match, if searchUp is enabled, then pop one segment off searchPath and search again.
+	for {
+		// Peek for module.
+		if accept&ActionableSearch_Module > 0 {
+			foundPath = filepath.Join(basisPath, remainingSearchPath, "module.wf")
+			m, e2 = ModuleFromFile(fsys, foundPath)
+			if serum.Code(e2) != wfapi.ECodeMissing { // notexist is ignored.
+				// Any error that's just just notexist: means our search has blind spots: error out.
+				err = wfapi.ErrorSearchingFilesystem("modules, plots, or formulas", e2)
+				return
+			}
+			// A module may also have a plot next to it; load that eagerly too.
+			p, e2 = PlotFromFile(fsys, filepath.Join(basisPath, remainingSearchPath, "plot.wf"))
+			if serum.Code(e2) != wfapi.ECodeMissing {
+				err = wfapi.ErrorSearchingFilesystem("loading plot associated with a module", e2)
+			}
+			return
+		}
+		// Peek for plot.
+		if accept&ActionableSearch_Plot > 0 {
+			foundPath = filepath.Join(basisPath, remainingSearchPath, "plot.wf")
+			p, e2 = PlotFromFile(fsys, foundPath)
+			if serum.Code(e2) != wfapi.ECodeMissing { // notexist is ignored.
+				// Any error that's just just notexist: means our search has blind spots: error out.
+				err = wfapi.ErrorSearchingFilesystem("modules, plots, or formulas", e2)
+				return
+			}
+		}
+
+		// Peek for Formula.
+		// TODO
```

Comment:
Without a description I assume this TODO was supposed to be implemented in this PR.

Reply:
It's not reachable where this function gets used by ferk mode.  It won't come up until we replace the current "run" subcommand's search logic with this function (that one does look for and use formulas, whereas ferk as yet has not supported that).  Or if we expanding the features of ferk to support formulas.

Probably worth turning into a panic, though.

Reply:
Okay. Definitely want a panic and/or additional comments.